### PR TITLE
add API to pass a filterRect to point cloud requests

### PR DESCRIPTION
**Pull request by uclaros**
## Description
This is preliminary API work for an upcoming PR that should allow limiting the 2D extents of 3D views.

### CODE REVIEW

File: src/core/pointcloud/qgspointcloudrequest.h [hunk 1]
```diff
@@ -48,8 +50,20 @@ class CORE_EXPORT QgsPointCloudRequest
     //! Set attributes filter in the request
     void setAttributes( const QgsPointCloudAttributeCollection &attributes );
 
+    /**
+     * Returns the rectangle from which points will be taken. If the returned rectangle is empty, then no filter rectangle is set.
+     * \since QGIS 3.30
+     */
+    QgsRectangle filterRect() const { return mExtent; }
```
Comment (nyalldawson) [inline src/core/pointcloud/qgspointcloudrequest.h:57]: Can you make a note here that the rect must be in the layer's crs?

Comment (uclaros) [reply src/core/pointcloud/qgspointcloudrequest.h:57]: done

File: src/core/pointcloud/qgseptdecoder.h [hunk 2]
```diff
@@ -19,27 +19,24 @@
...
+  QgsPointCloudBlock *decompressBinary( const QByteArray &data, const QgsPointCloudAttributeCollection &attributes, const QgsPointCloudAttributeCollection &requestedAttributes, const QgsVector3D &scale, const QgsVector3D &offset, QgsPointCloudExpression &filterExpression, QgsRectangle &filterRect );
+  QgsPointCloudBlock *decompressZStandard( const QString &filename, const QgsPointCloudAttributeCollection &attributes, const QgsPointCloudAttributeCollection &requestedAttributes, const QgsVector3D &scale, const QgsVector3D &offset, QgsPointCloudExpression &filterExpression, QgsRectangle &filterRect );
+  QgsPointCloudBlock *decompressZStandard( const QByteArray &data, const QgsPointCloudAttributeCollection &attributes, const QgsPointCloudAttributeCollection &requestedAttributes, const QgsVector3D &scale, const QgsVector3D &offset, QgsPointCloudExpression &filterExpression, QgsRectangle &filterRect );
```
Comment (wonder-sk) [inline src/core/pointcloud/qgseptdecoder.h:39]: would be good to make the filterRect reference const

File: src/core/pointcloud/qgslazdecoder.h [hunk 3]
```diff
@@ -86,10 +82,9 @@ class QgsLazDecoder
...
-    static QgsPointCloudBlock *decompressCopc( const QString &filename, QgsLazInfo &lazInfo, uint64_t blockOffset, uint64_t blockSize, int32_t pointCount, const QgsPointCloudAttributeCollection &requestedAttributes, QgsPointCloudExpression &filterExpression );
-    static QgsPointCloudBlock *decompressCopc( const QByteArray &data, QgsLazInfo &lazInfo, int32_t pointCount, const QgsPointCloudAttributeCollection &requestedAttributes, QgsPointCloudExpression &filterExpression );
+    static QgsPointCloudBlock *decompressLaz( const QString &filename, const QgsPointCloudAttributeCollection &requestedAttributes, QgsPointCloudExpression &filterExpression, QgsRectangle &filterRect );
+    static QgsPointCloudBlock *decompressLaz( const QByteArray &data, const QgsPointCloudAttributeCollection &requestedAttributes, QgsPointCloudExpression &filterExpression, QgsRectangle &filterRect );
+    static QgsPointCloudBlock *decompressCopc( const QByteArray &data, QgsLazInfo &lazInfo, int32_t pointCount, const QgsPointCloudAttributeCollection &requestedAttributes, QgsPointCloudExpression &filterExpression, QgsRectangle &filterRect );
```
Comment (wonder-sk) [inline src/core/pointcloud/qgslazdecoder.h:87]: here as well - let's use const ref for filterRect

File: src/core/pointcloud/qgseptdecoder.cpp [hunk 4]
```diff
@@ -99,37 +89,61 @@ QgsPointCloudBlock *_decompressBinary( const QByteArray &dataUncompressed, const
     return block.release();
   }
 
+  int xAttributeOffset, yAttributeOffset;
+  const QgsPointCloudAttribute *attributeX;
+  const QgsPointCloudAttribute *attributeY;
```
Comment (wonder-sk) [inline src/core/pointcloud/qgseptdecoder.cpp:94]: would be good to initialize the pointers to nullptr - just for sure :-)

File: src/core/pointcloud/qgslazdecoder.cpp [hunk 5]
```diff
@@ -519,9 +519,23 @@ QgsPointCloudBlock *__decompressLaz( FileType &file, const QgsPointCloudAttribut
       return block.release();
     }
 
+    int xAttributeOffset, yAttributeOffset;
+    const QgsPointCloudAttribute *attributeX;
+    const QgsPointCloudAttribute *attributeY;
```
Comment (wonder-sk) [inline src/core/pointcloud/qgslazdecoder.cpp:524]: also here and in decompressCopc - best to init the pointers

**Review by wonder-sk — COMMENTED** [review]
It looks good - just minor code comments...

File: src/core/pointcloud/qgseptdecoder.h (hunk 2 above)
Comment (uclaros) [reply src/core/pointcloud/qgseptdecoder.h:39]: We're modifying the filterRect to apply the node scale and offset to it. If we make it a const reference we would need to create a new scaled-offsetted QgsRectangle.

**Review by wonder-sk — APPROVED** [review]

**Pull request merged**